### ISSUE TITLE
rustdoc unexpected panic for GSL crate

### ISSUE THREAD

**Issue by jturner314** [rust-lang/rust#47728]
Calling `cargo doc` on the [GSL crate](https://github.com/GuillaumeGomez/rust-GSL/tree/9279d3300de495b55451f13ee5075e61851243e3) results in an unexpected compiler panic.

Steps to reproduce:

1. `git clone https://github.com/GuillaumeGomez/rust-GSL.git`
2. `cd rust-GSL`
3. `git checkout 9279d3300de495b55451f13ee5075e61851243e3`
4. `RUST_BACKTRACE=1 cargo doc`

The output of step 4 is:

```
    Updating registry `https://github.com/rust-lang/crates.io-index`
warning: path `/home/jim/projects/rust-GSL/src/rgsl.rs` was erroneously implicitly accepted for library `rgsl`,
please rename the file to `src/lib.rs` or set lib.path in Cargo.toml
   Compiling c_vec v1.0.12
 Documenting libc v0.2.36
   Compiling libc v0.2.36
   Compiling pkg-config v0.3.9
 Documenting c_vec v1.0.12
   Compiling GSL v0.4.31 (file:///home/jim/projects/rust-GSL)
 Documenting GSL v0.4.31 (file:///home/jim/projects/rust-GSL)
error: internal compiler error: unexpected panic

note: the compiler unexpectedly panicked. this is a bug.

note: we would appreciate a bug report: https://github.com/rust-lang/rust/blob/master/CONTRIBUTING.md#bug-reports

note: rustc 1.23.0 (766bd11c8 2018-01-01) running on x86_64-unknown-linux-gnu

note: run with `RUST_BACKTRACE=1` for a backtrace

thread 'rustc' panicked at 'byte index 1 is not a char boundary; it is inside '”' (bytes 0..3) of `”.
Communications`', /checkout/src/libcore/str/mod.rs:2235:4
note: Some details are omitted, run with `RUST_BACKTRACE=full` for a verbose backtrace.
stack backtrace:
   0: std::sys::imp::backtrace::tracing::imp::unwind_backtrace
             at /checkout/src/libstd/sys/unix/backtrace/tracing/gcc_s.rs:49
   1: std::sys_common::backtrace::_print
             at /checkout/src/libstd/sys_common/backtrace.rs:68
   2: std::panicking::default_hook::{{closure}}
             at /checkout/src/libstd/sys_common/backtrace.rs:57
             at /checkout/src/libstd/panicking.rs:381
   3: std::panicking::default_hook
             at /checkout/src/libstd/panicking.rs:391
   4: std::panicking::rust_panic_with_hook
             at /checkout/src/libstd/panicking.rs:577
   5: std::panicking::begin_panic
             at /checkout/src/libstd/panicking.rs:538
   6: std::panicking::begin_panic_fmt
             at /checkout/src/libstd/panicking.rs:522
   7: rust_begin_unwind
             at /checkout/src/libstd/panicking.rs:498
   8: core::panicking::panic_fmt
             at /checkout/src/libcore/panicking.rs:71
   9: core::str::slice_error_fail
             at /checkout/src/libcore/str/mod.rs:0
  10: <kuchiki::tree::NodeRef as html_diff::ToOutput>::name
             at /checkout/src/libcore/str/mod.rs:1988
             at /checkout/src/libcore/option.rs:370
             at /checkout/src/libcore/str/mod.rs:1988
             at /checkout/src/libcore/str/mod.rs:1735
             at /cargo/registry/src/github.com-1ecc6299db9ec823/html-diff-0.0.4/src/lib.rs:65
  11: _ZN9html_diff18ElementInformation3new17h269f42ee8374d90fE.llvm.2B544E5D
  12: _ZN9html_diff15go_through_tree17h78c3f03358efd8f5E.llvm.2B544E5D
             at /cargo/registry/src/github.com-1ecc6299db9ec823/html-diff-0.0.4/src/lib.rs:284
  13: _ZN9html_diff15go_through_tree17h78c3f03358efd8f5E.llvm.2B544E5D
             at /cargo/registry/src/github.com-1ecc6299db9ec823/html-diff-0.0.4/src/lib.rs:304
  14: _ZN9html_diff15go_through_tree17h78c3f03358efd8f5E.llvm.2B544E5D
             at /cargo/registry/src/github.com-1ecc6299db9ec823/html-diff-0.0.4/src/lib.rs:304
  15: _ZN9html_diff15go_through_tree17h78c3f03358efd8f5E.llvm.2B544E5D
             at /cargo/registry/src/github.com-1ecc6299db9ec823/html-diff-0.0.4/src/lib.rs:304
  16: html_diff::get_differences
             at /cargo/registry/src/github.com-1ecc6299db9ec823/html-diff-0.0.4/src/lib.rs:316
  17: _ZN7rustdoc4html6render15render_markdown17hfb2d0eb2ea37f0c2E.llvm.8ED23A91
             at /checkout/src/librustdoc/html/render.rs:1841
  18: _ZN7rustdoc4html6render13document_full17hd427ee8554321191E.llvm.8ED23A91
             at /checkout/src/librustdoc/html/render.rs:1897
  19: _ZN7rustdoc4html6render8document17ha48c00652eb54be9E.llvm.8ED23A91
             at /checkout/src/librustdoc/html/render.rs:1822
  20: <rustdoc::html::render::Item<'a> as core::fmt::Display>::fmt
             at /checkout/src/librustdoc/html/render.rs:1935
             at /checkout/src/librustdoc/html/render.rs:1765
  21: core::fmt::write
             at /checkout/src/libcore/fmt/mod.rs:1027
             at /checkout/src/libcore/fmt/mod.rs:995
  22: std::io::Write::write_fmt
             at /checkout/src/libstd/io/mod.rs:1144
  23: rustdoc::html::layout::render
             at /checkout/src/librustdoc/html/layout.rs:37
  24: _ZN7rustdoc4html6render7Context11render_item17hb6bbcfb238980096E.llvm.8ED23A91
  25: _ZN7rustdoc4html6render7Context4item17hb15a40f1b47433e8E.llvm.8ED23A91
             at /checkout/src/librustdoc/html/render.rs:1537
  26: rustdoc::html::render::run
             at /checkout/src/librustdoc/html/render.rs:1440

error: Could not document `GSL`.

Caused by:
  process didn't exit successfully: `rustdoc --crate-name rgsl src/rgsl.rs -o /home/jim/projects/rust-GSL/target/doc -L dependency=/home/jim/projects/rust-GSL/target/debug/deps --extern c_vec=/home/jim/projects/rust-GSL/target/debug/deps/libc_vec-394b5b1d81996265.so --extern c_vec=/home/jim/projects/rust-GSL/target/debug/deps/libc_vec-394b5b1d81996265.rlib --extern libc=/home/jim/projects/rust-GSL/target/debug/deps/liblibc-b1ca85687f9f2272.rlib --cfg feature="v2"` (exit code: 101)
```

On my machine, `rustc --version --verbose` yields

```
rustc 1.23.0 (766bd11c8 2018-01-01)
binary: rustc
commit-hash: 766bd11c8a3c019ca53febdcd77b2215379dd67d
commit-date: 2018-01-01
host: x86_64-unknown-linux-gnu
release: 1.23.0
LLVM version: 4.0
```

**Comment by devonhollowood** [comment]
It looks to me like this is caused by [this issue](https://github.com/GuillaumeGomez/html-diff-rs/issues/12) in `html-diff`. That issue was fixed in `html-diff 0.0.6`, which is what the nightly `rustdoc` uses (`rust 1.23.0` uses `html-diff 0.0.4`). To verify this, I tried out your steps-to-reproduce with the nightly compiler, and it compiled fine. So I think this is fixed now.

**Comment by jturner314** [comment]
I confirmed that this is fixed on nightly; thanks @devonhollowood. I'll try to remember to test on nightly before reporting issues next time.
